File: EasyNetDemo/UI/Photos/Controllers/PhotosTableViewController.m
//
//  PhotosTableViewController.m
//  EasyNetDemo
//
//  Created by Leo Lee on 2021/4/27.
//

#import <SVProgressHUD/SVProgressHUD.h>

#import "PhotosTableViewController.h"
#import "ProfileTableViewController.h"

#import "PhotosTableViewPresenter.h"
#import "PhotosTableViewDataSource.h"

#import "PhotosTableViewCell.h"

#import "NetworkListener.h"

@interface PhotosTableViewController () <PhotosTableView>

@property (nonatomic, strong) UITableView *tableView;
@property (nonatomic, strong) PhotosTableViewPresenter *presenter;
@property (nonatomic, strong) PhotosTableViewDataSource *dataSource;

@end

@implementation PhotosTableViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.navigationItem.title = @"Photos";
    
    [[NetworkListener sharedListener] networkListenerChangeStatusUsingBlock:^(NetworkListenerStatus status) {
        [self sendRequest];
    }];
    [self setupTableView];
}

- (void)traitCollectionDidChange:(UITraitCollection *)previousTraitCollection {
    [super traitCollectionDidChange:previousTraitCollection];
    
    if (@available(iOS 12.0, *)) {
        if (previousTraitCollection.userInterfaceStyle == UIUserInterfaceStyleDark) {
            [SVProgressHUD setDefaultStyle:SVProgressHUDStyleDark];
        } else {
            [SVProgressHUD setDefaultStyle:SVProgressHUDStyleLight];
        }
    } else {
        [SVProgressHUD setDefaultStyle:SVProgressHUDStyleDark];
    }
}

#pragma mark - Private

- (void)sendRequest {
    __weak typeof(self) weakSelf = self;
    [SVProgressHUD showWithStatus:nil];
    [self.presenter requestPhotosWithSuccess:^{
        [weakSelf.tableView reloadData];
        [SVProgressHUD showSuccessWithStatus:nil];
        [SVProgressHUD dismissWithDelay:0.2];
    } failure:^(NSError * _Nullable error) {
        [SVProgressHUD showErrorWithStatus:nil];
        [SVProgressHUD dismissWithDelay:0.2];
    }];
}

- (void)setupTableView {
    [self.view addSubview:self.tableView];
    [self.tableView registerClass:[PhotosTableViewCell class] forCellReuseIdentifier:NSStringFromClass([PhotosTableViewCell class])];
    [self.tableView.topAnchor constraintEqualToAnchor:self.view.topAnchor].active = YES;
    [self.tableView.leftAnchor constraintEqualToAnchor:self.view.leftAnchor].active = YES;
    [self.tableView.bottomAnchor constraintEqualToAnchor:self.view.bottomAnchor].active = YES;
    [self.tableView.rightAnchor constraintEqualToAnchor:self.view.rightAnchor].active = YES;
}

#pragma mark - PhotosTableView

- (void)photosTableViewPresenter:(PhotosTableViewPresenter *)presenter didSelectUserId:(NSNumber *)userId {
    ProfileTableViewController *viewController = [[ProfileTableViewController alloc] init];
    viewController.userId = userId;
    [self.navigationController pushViewController:viewController animated:YES];
}

#pragma mark - Getter

- (UITableView *)tableView {
    if (!_tableView) {
        _tableView = [[UITableView alloc] initWithFrame:CGRectZero style:UITableViewStylePlain];
        _tableView.translatesAutoresizingMaskIntoConstraints = NO;
        _tableView.tableHeaderView = [UIView new];
        _tableView.tableFooterView = [UIView new];
        _tableView.delegate = self.dataSource;
        _tableView.dataSource = self.dataSource;
    }
    return _tableView;
}

- (PhotosTableViewPresenter *)presenter {
    if (!_presenter) {
        _presenter = [[PhotosTableViewPresenter alloc] init];
        _presenter.view = self;
    }
    return _presenter;
}

- (PhotosTableViewDataSource *)dataSource {
    if (!_dataSource) {
        _dataSource = [[PhotosTableViewDataSource alloc] initWithPresenter:self.presenter];
    }
    return _dataSource;
}

@end

File: EasyNetDemoSwift/UI/PhotosTableViewController.swift
//
//  PhotosTableViewController.swift
//  EasyNetDemoSwift
//
//  Created by lidan on 2021/5/27.
//

import UIKit
import RxSwift
import RxCocoa
import SVProgressHUD

class PhotosTableViewController: UIViewController {
    
    private lazy var tableView: UITableView = {
        let tableView = UITableView(frame: .zero, style: .plain)
        tableView.translatesAutoresizingMaskIntoConstraints = false
        tableView.tableHeaderView = UIView()
        tableView.tableFooterView = UIView()
        tableView.rowHeight = CGFloat(120)
        tableView.register(PhotosTableViewCell.self, forCellReuseIdentifier: String(describing: PhotosTableViewCell.self))
        return tableView
    }()
    
    private let itemsObservable = BehaviorRelay<[PhotoModel]>(value: [])
    
    private let disposeBag = DisposeBag()

    override func viewDidLoad() {
        super.viewDidLoad()
        
        navigationItem.title = "Photos"
        view.backgroundColor = .white
        
        NetworkListener.shared.networkListenerChangeStatus { [weak self] _ in
            self?.sendRequest()
        }
        setupTableView()
    }
    
    override func traitCollectionDidChange(_ previousTraitCollection: UITraitCollection?) {
        super.traitCollectionDidChange(previousTraitCollection)
        
        if #available(iOS 12.0, *) {
            if previousTraitCollection?.userInterfaceStyle == .dark {
                SVProgressHUD.setDefaultStyle(.dark)
            } else {
                SVProgressHUD.setDefaultStyle(.light)
            }
        } else {
            SVProgressHUD.setDefaultStyle(.light)
        }
    }

}

extension PhotosTableViewController {
    
    private func setupTableView() {
        view.addSubview(tableView)
        
        NSLayoutConstraint.activate([
            tableView.topAnchor.constraint(equalTo: view.topAnchor),
            tableView.leftAnchor.constraint(equalTo: view.leftAnchor),
            tableView.bottomAnchor.constraint(equalTo: view.bottomAnchor),
            tableView.rightAnchor.constraint(equalTo: view.rightAnchor)
        ])
        
        tableView.rx.setDelegate(self).disposed(by: disposeBag)
        
        itemsObservable.bind(to: tableView.rx.items(cellIdentifier: String(describing: PhotosTableViewCell.self), cellType: PhotosTableViewCell.self)
        ) { row, photo, cell in
            cell.photoModel = photo
            cell.textLabel?.text = photo.title
        }.disposed(by: disposeBag)
        
        tableView.rx.itemSelected.bind { [weak self] indexPath in
            let profile = ProfileTableViewController()
            profile.userId = self?.itemsObservable.value[indexPath.row].userId ?? 0.0
            self?.navigationController?.pushViewController(profile, animated: true)
        }.disposed(by: disposeBag)
        
    }
    
    private func sendRequest() {
        SVProgressHUD.show(withStatus: nil)
        NetworkClient.default.requestPhotos { request in
            SVProgressHUD.showSuccess(withStatus: nil)
            SVProgressHUD.dismiss(withDelay: 0.2)
            if let responses = request.convertObject as? [PhotoResponse] {
                self.itemsObservable.accept(responses.map { PhotoModel(photoResponse: $0) })
            }
        } failure: { request in
            SVProgressHUD.showSuccess(withStatus: nil)
            SVProgressHUD.dismiss(withDelay: 0.2)
            print(request.error ?? "")
        }
    }
    
}


extension PhotosTableViewController: UITableViewDelegate {
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
    }
}

